js/internal/modules/modules.go

```diff
@@ -35,7 +35,17 @@ var (
 func Get(name string) interface{} {
 	mx.RLock()
 	defer mx.RUnlock()
-	return modules[name]
+	mod := modules[name]
+	if i, ok := mod.(PerTestModule); ok {
+		return i.NewVUModule()
+	}
+	return mod
+}
+
+// PerTestModule is a simple interface representing a per test module object that can be used to
+// make a per VU instance of the Module
+type PerTestModule interface {
```

Comment:
The idea is that this interface should be the per-test instance and it generates a per VU module. Which it currently does. I do agree it is currently mostly per instance one, but making it truly per test will take more refactoring of the modules so I am not making this in this PR. So I would argue the name is pretty accurate both in what it can be used (in production code) and in what it will stay in the future(even though then there will be a one more level).

The idea is that I will actually continue with #1858  and add the code so that all of that is implemented, but doing it in this PR seems unnecessary and this is all in the internal module, so nobody can use it beside us (mostly xk6 I guess ;)).

Reply:
I get the ideas and I still maintain this current name is misleading. My suggestions above aren't great, but are definitely less confusing. And if not them, how about `HasPerVUModules`?

btw do we really need a name for an interface that has per-instance modules? Wouldn't that be covered by just having a different constructor? For example, this code: https://github.com/loadimpact/k6/blob/82cab8b693963659d5f0cfb4cbf785c792577a0f/js/modules/k6/data/data.go#L36-L38

can be something like this:
```go
modules.Register("k6/data", &data{
    sharedObjects: common.NewSharedObjects(),
    // ... whatever other per-instance state we want
}) 
```

Reply:
I am for just leaving it for the PR I have already started and will probably publish next week or so. 

The answer the other constructor variant:
yes ... but only if you don't want for you to be able to juts `import SharedArray from "k6/data"` in the js packages without needing to work with `data` (which will need to be exported) from the `k6/data` module directly. 

So as always , yes we can remove some abstraction, but we have to to then know how to make it  work everywhere. And I would argue this will be even more useful for `xk6` extensions as ... I technically could've implemented this with one more `SharedArray` hack like this https://github.com/loadimpact/k6/blob/82cab8b693963659d5f0cfb4cbf785c792577a0f/js/common/initenv.go#L70-L73

Reply:
> I am for just leaving it for the PR I have already started and will probably publish next week or so.

:confused: By then v0.31.0 will be released, so why should this have a misleading name until then when the fix is just to rename it?

Reply:
but your renames are ... more inaccurate ... 
p.s. this is why I wanted to leave it unnamed

Reply:
cc @imiric

Reply:
Again the current interface is meant for a *per test* (which currently is per instance mostly, but there is no difference in production code so :shrug: ) that has a way to create a *per vu* module. 

So if anything the name should `PerInstanceModule` not `PerTestModule`, but it is definitely not *per VU* ...

Reply:
> Again the current interface is meant for a per test (...) that has a way to create a per vu module.

Can I have a per-test module that doesn't implement `PerTestModule`? Of course. Then `PerTestModule` is not a good name for the interface, is it?

Leaving interfaces unnamed is just postponing the problem where we apparently don't agree on naming things, which probably means we don't think of this in the same terms, which probably means it's a good idea to discuss them now and reach a consensus...

> but it is definitely not per VU
That's why I suggested `HasPerVUModules` as a name of the interface. It signals that the parent object, which is used in `modules.Register()`, has a unique instance of the module unique in every VU. Here's another suggestion:
```go
// HasModuleInstancePerVU should be implemented by all native Golang modules that 
// would require per-VU state. k6 will call their NewModuleInstance() methods
// every time a VU imports the module and use its result as the returned object.
type HasModuleInstancePerVU interface {
    NewModuleInstance() interface{}
}

Reply:
the point of this code is to make the current responseCallback possible without:
1. one more global but actually per module shared state store
2. not rewriting all the modules to do it.

So it is only checking for a specific case and that will be changed and *all* modules will just implement the interface as that takes 5 lines and removes the wondering of whetether or not your module is correct or not. And then I would expect that we will just make it required for xk6 modules as well after some time of just logging that the should be implementing the new interface ... 

For the purposes of actually merging this code .. .I am fine with w/e I just don't think that a temporary name for a temporary interface in a an *internal* module, matters enough for a design committee

Reply:
It matters because I am the one reviewing the code and the current name doesn't make sense to me. The fact that we still disagree about this means that either the name is bad, or I am completely missing something (which, unless I am a complete idiot, still means something in the code can be better and more obvious). 

In any case, even an _internal_ name has to make sense and describe accurately the thing it's naming... I _think_ I understand how, what and why and  you're trying to achieve here, but I might be wrong, so I'll take a pause and re-read through this whole discussion and https://github.com/loadimpact/k6/issues/1858 later today.

Reply:
I agree that the current `PerTestModule` name is slightly confusing. Considering that the function is called `NewVUModule()`, `HasPerVUModule` makes sense to me.

In any case, it's an internal interface we can always rename later, so don't think we should spend a lot of time on this.

Reply:
I've laid out my arguments in https://github.com/loadimpact/k6/issues/1858#issuecomment-788867806 how I think this should be evolved in k6 v0.32.0 and why. Having thought about this a lot more and re-read the discussions here and in the issue, I still maintain that this is a better name than the current interface:
```go
type HasModuleInstancePerVU interface {
    NewModuleInstance() interface{}
}
```
Maybe even this, if we want to be super-descriptive:
```go
type HasModuleInstancePerVU interface {
    NewModuleInstancePerVU() interface{}
}
```